src/Build/Logging/BinaryLogger/ProjectImportsCollector.cs

```diff
@@ -55,33 +36,46 @@ public byte[] GetAllBytes()
 
         public ProjectImportsCollector(string logFilePath, bool createFile, string sourcesArchiveExtension = ".ProjectImports.zip")
         {
-            try
+            if (createFile)
             {
-                if (createFile)
-                {
-                    ArchiveFilePath = Path.ChangeExtension(logFilePath, sourcesArchiveExtension);
-                    _stream = new FileStream(ArchiveFilePath, FileMode.Create, FileAccess.ReadWrite, FileShare.Delete);
-                }
-                else
+                // Archive file will be stored alongside the binlog
+                ArchiveFilePath = Path.ChangeExtension(logFilePath, sourcesArchiveExtension);
+            }
+            else
+            {
+                string cacheDirectory = FileUtilities.GetCacheDirectory();
+                if (!Directory.Exists(cacheDirectory))
                 {
-                    _stream = new MemoryStream();
+                    Directory.CreateDirectory(cacheDirectory);
                 }
-                _zipArchive = new ZipArchive(_stream, ZipArchiveMode.Create, true);
+
+                // Archive file will be temporarily stored in MSBuild cache folder and deleted when no longer needed
+                ArchiveFilePath = Path.Combine(
+                    cacheDirectory,
```

Comment:
This folder is one that is already secured to the current user on Unix?

Reply:
Yes, it's defined as

https://github.com/dotnet/msbuild/blob/39e20dce0c00a572f1fbd20a0a520d4e941bbc33/src/Shared/FileUtilities.cs#L131

calling our hardened

https://github.com/dotnet/msbuild/blob/39e20dce0c00a572f1fbd20a0a520d4e941bbc33/src/Shared/TempFileUtilities.cs#L37-L68